torch/csrc/api/include/torch/optim/serialize.h

```diff
@@ -40,6 +40,101 @@ void serialize(
+void serialize(
+    serialize::OutputArchive& archive,
+    const std::string& key,
+    const ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>>& state) {
+
+  serialize::OutputArchive state_archive;
+  for (const auto& item : state) {
+    serialize::OutputArchive param_state_archive; // For each OptimizerParamState
+    std::string tensorimpl_key = item.first;
+    const DerivedOptimizerParamState& curr_state = static_cast<const DerivedOptimizerParamState&>(*(item.second.get()));
+    curr_state.serialize(param_state_archive);
+    state_archive.write(tensorimpl_key, param_state_archive);
+  }
+  archive.write(key, state_archive);
+}
+
+template <typename DerivedOptimizerParamState>
+void serialize(
+    serialize::InputArchive& archive,
+    const std::string& key,
+    ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>>& state) {
+
+  serialize::InputArchive state_archive;
+  archive.read(key, state_archive);
+  std::vector<std::string> tensorimpl_keys = state_archive.keys();
+  for (const std::string& tensorimpl_key : tensorimpl_keys) {
+    serialize::InputArchive param_state_archive;
+    archive.read(tensorimpl_key, param_state_archive);
+    DerivedOptimizerParamState param_state;
+    param_state.serialize(param_state_archive);
+    state[tensorimpl_key] = c10::guts::make_unique<DerivedOptimizerParamState>(param_state);
+  }
+}
+
+template <typename DerivedOptimizerParamOptions>
+void serialize(
+    serialize::OutputArchive& archive,
+    const std::string& key,
+    const std::vector<OptimizerParamGroup>& param_groups) {
+
+  serialize::OutputArchive param_groups_archive;
+  for (size_t i = 0; i < param_groups.size(); i++) {
+    serialize::OutputArchive param_group_archive; // For each OptimizerParamState
+    std::vector<Tensor> params = param_groups[i].params();
+    param_group_archive.write(
+        "params/size", torch::tensor(static_cast<int64_t>(params.size())));
+    for(size_t index = 0; index < params.size(); index++) {
+      param_group_archive.write(
+          "params/" + c10::guts::to_string(index), IValue(c10::guts::to_string(params[index].unsafeGetTensorImpl())));
+    }
+
+    const DerivedOptimizerParamOptions& param_group_options = static_cast<const DerivedOptimizerParamOptions&>(param_groups[i].options());
+
+    serialize::OutputArchive param_group_options_archive;
+    param_group_options.serialize(param_group_options_archive);
+    param_group_archive.write("options", param_group_options_archive);
+    param_groups_archive.write(key + "/" + c10::guts::to_string(i), param_group_archive);
+  }
+  //serialize param_groups "params"->vector<string>, "options" -> options
+  archive.write(key, param_groups_archive);
+}
+
+template <typename DerivedOptimizerParamOptions>
+void serialize(
+    serialize::InputArchive& archive,
+    const std::string& key,
+    std::vector<std::pair<std::vector<std::string>, OptimizerOptions>>& param_groups) {
```

Comment:
I remember we discussed whether we should use `std::vector<std::pair<std::vector<std::string>, OptimizerOptions>>` type to represent the `param_groups` in this function, and at that time I thought that it was only used as an internal variable in the function, not as a parameter in the function signature. Using it as a parameter in the function signature has the following issue:

This `torch::optim::serialize<DerivedOptimizerParamOptions>(...)` function can be considered as public API since it can be used by external users. Usually we want to provide strong type coupling between our public APIs, in the sense that since `OptimizerBase`'s `param_groups_` is of type `std::vector<OptimizerParamGroup>`, this function should take in `std::vector<OptimizerParamGroup>` as well, otherwise the user would not be able to directly pass in `OptimizerBase::param_groups_` to this function and get the deserialized param groups back, and they have to do a conversion which adds UX friction.

Reply:
There is another issue with using `OptimizerOptions` instead of `std::unique_ptr<OptimizerOptions>`, which is that object slicing (https://en.wikipedia.org/wiki/Object_slicing) would be performed when we do `param_groups.push_back(std::make_pair(params, param_group_options))`, which slices off all subclass information of `param_group_options` when it is being stored into the pair. To get around this problem we would always need to use `std::unique_ptr<OptimizerOptions>` instead.

Reply:
It might be worth checking whether the object slicing problem causes "key not found" error :)

Reply:
synced offline:  std::vector<std::pair<std::vector<std::string>, std::unique_ptr<OptimizerOptions>>> needs to be used because we are saving std::vector<std::string> and not std::vector<std::Tensor>. However such an API shouldn't be exposed to users so I have added a new function that takes in Optimizer class object as input and updates the state and param_groups inside the function as required